### DISCUSSION TITLE
Allow us to mutate the session on the client side without making any other call like getSession()

### DISCUSSION THREAD

**Discussion by phileasmah** [nextauthjs/next-auth#2267] [Ideas]
## Summary 💭

Include a ```setState``` to mutate the ```session``` state when using the ```useSession()``` hook. 

## Description 📓
In my app, I'm trying to build a settings page for profiles, like changing the display name etc, but I found that after changing the display name (name in database has been changed), the ```session``` state in ```useSession()``` stayed the same.

So, I was trying to find a way to mutate the session object of the current user and stumbled upon ```getSession()``` which (correct me if I'm wrong) is suppose to get a new session and replace it with the current one but according to this [`issue`](https://github.com/nextauthjs/next-auth/issues/596) it doe not work as intended. 

I was looking at the source code and thought wouldn't it be easier to just allow us to mutate the session state directly? (or are there security concerns I'm not aware of? In which case, it would be interesting to know). Looking at the code, it should be pretty easy to implement.

An advantage when compared to ```getSession()``` would obviously be that 1 less call needs to be made to the database, but again, I'm not sure if I am glossing over any security concerns etc

**Comment by balazsorban44** [comment]
Hi, and thanks for opening this. So when the session value changes and is being persisted either in a database or in the HttpOnly session cookie as JWT, the value can only be written server-side, so a trip to the server is inevitable. But I think we could expose a new function, something like `updateSession`, or better:

```js
const { data: session, update } = useSession()
...
const handleUpdate = async () => {
  update((prev) => {...prev: foo: "new session value"})
}
```
`update()` could take either a new session object, or a callback, where the argument is the previous session for convenience (mimicking the `setState` function of `useState`), that would not only update the session and persist it server-side, but would also update the session
in the client-side cache.

I am currently refactoring the `useSession()` API in #2236, that would allow us to easily return other properties from `useSession`

**Reply by balazsorban44** [reply]
I think it is a good idea, and I created an issue to track this as a feature request instead! Check out #2269, let's continue the discussion there
